### PR TITLE
magento/magento2#: Captcha. Improvement. Replace deprecated addError with addErrorMessage.

### PR DESCRIPTION
**Pull request by atwixfirster**
### Description (*)

Pull request replaces deprecated `addError` method with a new one - `addErrorMessage` in the `Magento_Captcha` module.

### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Thank you!

### CODE REVIEW

**Comment by southerncomputer** [comment]
Why not replace all AddSuccess, AddError, and AddException calls to new format?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @atwixfirster,
Could you review and fix failing static tests?

File: app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php [hunk 1]
```diff
@@ -143,10 +155,10 @@ public function execute(\Magento\Framework\Event\Observer $observer)
                 try {
                     $customer = $this->getCustomerRepository()->get($login);
                     $this->getAuthentication()->processAuthenticationFailure($customer->getId());
-                } catch (NoSuchEntityException $e) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php:146]: Hi,
This change is not related to this PR, could you revert this part as it was earlier?
If static tests complains about it - please add ignore for this case

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php (hunk 1 above)
Comment (atwixfirster) [reply app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php:146]: Hi, @ihor-sviziev! Thank you for notice!

Do you know the ignore command for this particular case? Static test error was something like "Empty catch statement..."

Thank you!

Comment (ihor-sviziev) [reply app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php:146]: Try to use following:
https://github.com/magento/magento2/blob/ee6bc3729e85ed1f1b5cd7c0eb9a3800bdd274c5/dev/tests/integration/testsuite/Magento/SalesRule/_files/cart_rule_100_percent_off_rollback.php#L23-L30

Comment (atwixfirster) [reply app/code/Magento/Captcha/Observer/CheckUserLoginObserver.php:146]: > Try to use following:

tried

Let's see if your magic can help here, @ihor-sviziev

Thank you!

**Comment by ihor-sviziev** [comment]
@magento run all

**Comment by sidolov** [comment]
@magento combine 24334 24284

**Comment by magento-engcom-team** [comment]
Hi @sidolov. Thank you for your request. I'm working on combining the pull requests for you

**Comment by magento-engcom-team** [comment]
@atwixfirster all pull requests have been successfully combined together: 
 - merged https://github.com/magento/magento2/pull/24334: magento/magento2#: Replace deprecated Magento\Customer\Model\Customer::isConfirmationRequired method
 - merged https://github.com/magento/magento2/pull/24284: magento/magento2#: Replace deprecated addError, addSuccess, addException methods in Magento/Customer/Controller/Account/Confirmation.php

**Review by dmytro-ch — CHANGES_REQUESTED** [review]
@atwixfirster , could you please also check the issue mentioned id the "child" PR #24334.
Thank you!

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests build

**Comment by ihor-sviziev** [comment]
@magento Sample Data Tests build

**Comment by ihor-sviziev** [comment]
@magento run Sample Data Tests build

**Comment by ihor-sviziev** [comment]
@magento run Static Tests build

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests build

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests build

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Customer/Model/Customer.php [hunk 2]
```diff
@@ -486,7 +492,8 @@ public function getAddressById($addressId)
      * Getting customer address object from collection by identifier
      *
      * @param int $addressId
-     * @return Address
+     * @return \Magento\Framework\DataObject
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Customer.php]: Isn’t it returns Address object? Looks like incorrect change

File: app/code/Magento/Customer/Model/Customer.php [hunk 3]
```diff
@@ -537,7 +545,8 @@ public function getAddresses()
     /**
      * Retrieve all customer attributes
      *
-     * @return Attribute[]
+     * @return array
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Customer.php]: It’s better to keep original return statement there. Or it was incorrect?

File: app/code/Magento/Customer/Model/ResourceModel/Customer.php [hunk 4]
```diff
@@ -21,9 +22,9 @@
 class Customer extends \Magento\Eav\Model\Entity\VersionControl\AbstractEntity
 {
     /**
-     * @var \Magento\Framework\Validator\Factory
+     * @var \Magento\Framework\Stdlib\DateTime
      */
-    protected $_validatorFactory;
+    protected $dateTime;
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/ResourceModel/Customer.php]: Why did you moved it above?

File: app/code/Magento/Customer/Model/Session.php [hunk 5]
```diff
@@ -175,9 +182,11 @@ public function __construct(
             $cookieMetadataFactory,
             $appState
         );
+        $this->_eventManager->dispatch('customer_session_init', ['customer_session' => $this]);
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Session.php]: Why you move it above?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @atwixfirster,
Could you review my comments?
Also there is some issues with tests for this PR, I think we can ignore these issues

File: app/code/Magento/Customer/Model/ResourceModel/Customer.php (hunk 4 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/ResourceModel/Customer.php]: Sort in alphabetical order ;) I can revert this change if you request it. Thanks

File: app/code/Magento/Customer/Model/Session.php (hunk 5 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php]: Sort in alphabetical order ;) 

I was wrong in this case... Thank you for notice

File: app/code/Magento/Customer/Model/ResourceModel/Customer.php (hunk 4 above)
Comment (ihor-sviziev) [reply app/code/Magento/Customer/Model/ResourceModel/Customer.php]: It’s better to keep only needed changes in order to prevent conflicts, so please revert it.

File: app/code/Magento/Customer/Model/Session.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Customer/Model/Session.php]: It’s better to keep only needed changes in order to prevent conflicts, so please revert it.

File: app/code/Magento/Customer/Model/Customer.php (hunk 3 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Customer.php]: Yes, the origin return statement is correct. 

Thanks, @ihor-sviziev

File: app/code/Magento/Customer/Model/Customer.php (hunk 2 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Customer.php]: Yes, it returns Address object.

File: app/code/Magento/Customer/Model/Customer.php [hunk 6]
```diff
@@ -485,8 +491,9 @@ public function getAddressById($addressId)
     /**
      * Getting customer address object from collection by identifier
      *
-     * @param int $addressId
+     * @param $addressId
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Customer.php]: Looks like you removed type of this parameter by mistake, please revert this change

File: app/code/Magento/Customer/Model/Session.php [hunk 7]
```diff
@@ -152,18 +157,20 @@ public function __construct(
...
         CustomerRepositoryInterface $customerRepository,
         GroupManagementInterface $groupManagement,
-        \Magento\Framework\App\Response\Http $response
+        \Magento\Framework\App\Response\Http $response,
+        AccountConfirmation $accountConfirmation = null
     ) {
-        $this->_coreUrl = $coreUrl;
-        $this->_customerUrl = $customerUrl;
         $this->_configShare = $configShare;
+        $this->_coreUrl = $coreUrl;
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Session.php:163]: Could you revert all not related changes? Originally it was sorted in the same way as constructor arguments, it was logically correct

Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php:163]: **done**

File: app/code/Magento/Customer/Model/Customer.php (hunk 6 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Customer.php]: **done**

File: app/code/Magento/Customer/Model/Session.php (hunk 5 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php]: **done**

File: app/code/Magento/Customer/Model/ResourceModel/Customer.php (hunk 4 above)
Comment (atwixfirster) [reply app/code/Magento/Customer/Model/ResourceModel/Customer.php]: yes. Now I understand this. Thank you

File: app/code/Magento/Customer/Model/Session.php [hunk 8]
```diff
@@ -354,10 +370,11 @@ public function setCustomerGroupId($id)
     }
 
     /**
-     * Get customer group id
-     * If customer is not logged in system, 'not logged in' group id will be returned
+     * Get customer group id. If customer is not logged in system, 'not logged in' group id will be returned.
      *
-     * @return int
+     * @return int|null
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Session.php]: I reviewed and didn't found case when there could be null. Could you point me?

Sorry, that's the latest comment, everything else looks good

Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php]: > Could you point me?

I can not point you to this particular case but PHPStorm points to app/code/Magento/Customer/Api/Data/CustomerInterface.php, [line 61](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Customer/Api/Data/CustomerInterface.php#L61) ;)

Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php]: **fixed**

Comment (atwixfirster) [reply app/code/Magento/Customer/Model/Session.php]: Thanks, @ihor-sviziev

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5753](https://jira.corp.magento.com/browse/ENGCOM-5753) has been created to process this Pull Request

File: app/code/Magento/Customer/Controller/Account/Confirmation.php [hunk 9]
```diff
@@ -1,21 +1,26 @@
...
  * See COPYING.txt for license details.
  */
 namespace Magento\Customer\Controller\Account;
 
+use Magento\Customer\Api\AccountManagementInterface;
+use Magento\Customer\Controller\AbstractAccount;
+use Magento\Customer\Model\Session;
 use Magento\Customer\Model\Url;
+use Magento\Framework\App\Action\HttpGetActionInterface as HttpGetActionInterface;
+use Magento\Framework\App\Action\HttpPosttActionInterface as HttpPostActionInterface;
```
Comment (engcom-Alfa) [inline app/code/Magento/Customer/Controller/Account/Confirmation.php]: You made spelling mistake in `HttpPosttActionInterface` . Please, delete the extra letter `t`

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
You made spelling mistake in `HttpPosttActionInterface` . Please, delete the extra letter `t`

File: app/code/Magento/Customer/Controller/Account/Confirmation.php (hunk 9 above)
Comment (ihor-sviziev) [reply app/code/Magento/Customer/Controller/Account/Confirmation.php]: Good catch 👍

Comment (atwixfirster) [reply app/code/Magento/Customer/Controller/Account/Confirmation.php]: Good catch, @engcom-Alfa !

Thank you!

**Comment by atwixfirster** [comment]
> You made spelling mistake in `HttpPosttActionInterface` . Please, delete the extra letter `t`

**fixed**

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5753](https://jira.corp.magento.com/browse/ENGCOM-5753) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed
